routing/router.go

```diff
@@ -123,7 +123,8 @@ type PaymentAttemptDispatcher interface {
 	// longer be in flight.  The switch shutting down is signaled by
 	// closing the channel. If the attemptID is unknown,
 	// ErrPaymentIDNotFound will be returned.
-	GetAttemptResult(attemptID uint64, paymentHash lntypes.Hash,
+	GetAttemptResult(attempt *channeldb.HTLCAttempt,
```

Comment:
Not sure I followed - if you wanna create onions then you should use `SendToOnions`? 

> Is the optimization on sphinx circuit creation needed to fix the bug? Do we gain much by way performance or more logical code separation by burying the decryption of errors within the Switch?

Nope it's not needed but an improvement - an easy way to measure it is to run the payment benchmark, which increases a bit. This is more of a logical code separation since there's no need to re-create the error decryptor if we know it's not a fail but a settle. 

> One argument for keeping the error decryption logic in the ChannelRouter could be that it does create the onion after all so will still need to handle or know about the hops + session key.

I don't think that's the case, all `paymentLifecycle` needs is to call `RequestRoute` and pass the route to the `htlcswitch`.

Anyway I removed this change to limit the scope of this PR, think we'll take a look at it again when working on #8834.